### ISSUE TITLE
sometimes I want to have some constant values that are part of a set

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4671]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#4671|
|Type|enhancement|
|Created|2010-10-05 15:10:16Z|
|Branch|https://github.com/twisted/twisted/tree/constants-4671|

Twisted contains several piles of named constants that represent interesting constant values that belong to a set.  Some examples include: HTTP methods, HTTP response codes, SMTP response codes, OSCAR SNAC codes, AMP special keys (ASK, ANSWER, COMMAND), IRC RPL codes, sftp file flags, and telnet negotiation options.

These are all represented in idiosyncratic and incompatible ways.  Many of them are just integers bound to names.  Many of them need some kind of bi-directional mapping between a string and an integer, or some associated descriptive text.  In some cases we automate that mapping construction, in others we do it by hand.

I'd like to have a single idiom for representing this type of value, one which repr()s to the FQPN of the code's definition (or something similar) so it's easy to look up when debugging, but also contains the information relevant to the protocol (the numeric code, the bytes to send, etc) as a documented attribute.

For some of these, for example telnet negotiation options and conch's `FXF_*` flags, it may also be helpful to define some kind of "bitwise" operations, so that "FXF_WRITE | FXF_CREAT | FXF_TRUNC" actually looks like something meaningful, and not just '26', when printed.  Of course, there should also be an easy way to get 26 (or should I say, 0x1A) out of that combined value as well as an individual one.  For example, if `FXF_WRITE.asInt() → 2`, then `(FXF_WRITE | FXF_CREAT | FXF_TRUNC).asInt() → 26`.  (And no, we don't necessarily need to use '|' and '&' but it may make sense to do so.)

Some other useful functionality: there should be an object representing the total set of statuses or codes or flags, so that it's easy to ask, "is this a value I understand", or, "I have an integer / some bytes / a byte, please give me the associated status object for that value".

This ticket is to implement the library which these different use-cases would use.  It is probably not a good idea to try to do this in isolation, since it would be hard to get right without lots of specific reference to use-cases, but rather, to take the opportunity to implement this when some maintenance on such codes comes up.

<details><summary>Searchable metadata</summary>

```
trac-id__4671 4671
type__enhancement enhancement
reporter__glyph glyph
priority__low low
milestone__ 
branch__branches_constants_4671 branches/constants-4671
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1286291416000000 1286291416000000
changetime__1331456720000000 1331456720000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Barry apparently wrote something in this vein, we should probably look at it: https://launchpad.net/flufl.enum flufl.enum].

Or perhaps he could be convinced to relicense more permissively.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [32533]) Branching to 'constants-4671'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

The scope of this ticket is a little wide.  There are several similar features here, but they can be implemented separately, and will be used separately:

  1. [ticket:5382 Plain valueless named constants, part of a particular collection of such constants]
  1. [ticket:5383 Constants with values assigned from monotonically increasing integer sequence]
  1. [ticket:5384 Constants with values assigned corresponding to bits such that they can be combined]

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [33677]) Merge flag-constants-5384 

Author: exarkun
Reviewer: therve
Fixes: [#5384](https://github.com/twisted/twisted/issues/5384), [#4671](https://github.com/twisted/twisted/issues/4671)

Add bitvector-like constant support to `twisted.python.constants`.
